src/openrct2-ui/windows/LoadSave.cpp

```diff
@@ -161,17 +161,85 @@ static void window_loadsave_sort_list();
 
 static rct_window* window_overwrite_prompt_open(const char* name, const char* path);
 
-void window_loadsave_set_loadsave_callback(loadsave_callback cb)
+static utf8* getLastDirectoryByType(int32_t type)
 {
-    _loadSaveCallback = cb;
+    switch (type & 0x0E)
+    {
+        case LOADSAVETYPE_GAME:
+            return gConfigGeneral.last_save_game_directory;
+
+        case LOADSAVETYPE_LANDSCAPE:
+            return gConfigGeneral.last_save_landscape_directory;
+
+        case LOADSAVETYPE_SCENARIO:
+            return gConfigGeneral.last_save_scenario_directory;
+
+        case LOADSAVETYPE_TRACK:
+            return gConfigGeneral.last_save_track_directory;
+
+        default:
+            return nullptr;
+    }
+}
+
+static void getInitialDirectoryByType(const int32_t type, char* path, size_t pathSize)
+{
+    const char* subdir = nullptr;
+    switch (type & 0x0E)
+    {
+        case LOADSAVETYPE_GAME:
+            subdir = "save";
+            break;
+
+        case LOADSAVETYPE_LANDSCAPE:
+            subdir = "landscape";
+            break;
+
+        case LOADSAVETYPE_SCENARIO:
+            subdir = "scenario";
+            break;
+
+        case LOADSAVETYPE_TRACK:
+            subdir = "track";
+            break;
+    }
+
+    platform_get_user_directory(path, subdir, pathSize);
 }
 
-static int32_t window_loadsave_get_dir(utf8* last_save, char* path, const char* subdir, size_t pathSize)
+static const char* getFilterPatternByType(const int32_t type, const bool isSave)
 {
+    switch (type & 0x0E)
+    {
+        case LOADSAVETYPE_GAME:
+            return isSave ? "*.sv6" : "*.sv6;*.sc6;*.sv4;*.sc4";
+
+        case LOADSAVETYPE_LANDSCAPE:
+            return isSave ? "*.sc6" : "*.sc6;*.sv6;*.sc4;*.sv4";
+
+        case LOADSAVETYPE_SCENARIO:
+            return "*.sc6";
+
+        case LOADSAVETYPE_TRACK:
+            return isSave ? "*.td6" : "*.td6;*.td4";
+
+        case LOADSAVETYPE_IMAGE:
```

Comment:
Not very related to this PR, buw while you're at refactoring anyway, could you rename this to `LOADSAVETYE_HEIGHTMAP` and in the function above `getInitialDirectoryByType` set the heightmap directory as the default? Loading of images is not being used by anything else yet, and heightmaps could have their own default directory.